docs/extended-collection-types.md

```diff
@@ -0,0 +1,169 @@
+
+## Design
+
+The current `list()` type will remain unchanged for BYOND compatibility, with the other types being added *in addition to* these legacy lists.
+
+The following collection types will essentially be exposed directly from C# to DM through a minimal wrapper API:
+
+- **Array**: A fixed-length collection of elements.
+- **Dictionary**: A collection of key-value pairs. Equivalent to existing associative lists, and will serve as the internal implementation of BYOND's planned `alist()` feature.
+- **HashSet**: A collection that contains no duplicate elements.
+- **ODList**: A dynamically resizable collection that is purely a list with none of the assoc-list functionality that BYOND lists have.
+- **Queue**: A first-in first-out (FIFO) collection.
+- **Stack**: A last-in first-out (LIFO) collection.
+
+Generally speaking, the helper procs from `list()` will also operate on other collections where applicable (e.g. `list.Join()` and `array.Join()` would be equivalent). A few collections will implement additional helpers as noted below.
+
+**Note:** Initializing OpenDream's collection types with a fixed size, for example `var/array/foo[10]`, will initialize it with a `capacity` of 10 but *will not* insert any `null` elements; the actual contents of the collection will be empty. This is **breaking behavior** from BYOND lists such as `var/list/L[10]`, which inserts 10 `null` elements into the list. BYOND lists in OpenDream will preserve this behavior for compatibility. Otherwise, see the "Collection Capacity API" subsection below for more information.
+
+### Array
+
+A fixed-length collection of elements. Almost identical to `list()` except it cannot be resized.
+
+Example usage:
+```
+var/array/foo = array("a","b","c")
+foo += "z" // Compile-time error, cannot resize arrays
+
+world.log << foo.Join() // Prints "abc"
+
+foo[1] = "z" // Valid modification
+```
+
+### Dictionary
+A collection of key-value pairs. Equivalent to existing associative lists, and will serve as the internal implementation of BYOND's planned `alist()` feature.
+
+Example usage:
+```
+var/dict/bar = array("a" = 1,"b" = 2,"c" = 3)
+bar += "d" // Compile-time error, no value specified
+
+bad["d"] = 4 // Valid modification
+
+world.log << bar["c"] // Prints 3
+
+world.log << bar["test"] // Runtime error, key not found
+
+bar.contains_key("test") // Returns FALSE
+
+bar.get_value("test") // Returns null, since the key doesn't exist
+
+world.log << bar[1] // Runtime error, dicts can't be numerically indexed
+```
+
+`dict.keys` and `dict.values` will return a new `array()` of the dictionary's keys and values, respectively. Modifying these will not modify the existing dictionary.
```

Comment:
If its not going to be linked, maybe it should return a constant list instead.

Reply:
I have no strong opinion so I'll see if anyone else weighs in.

Reply:
It might be better to define these as being some kind of iterator so that the implementation of keys/values can be allocation-free.